filter.py

```diff
@@ -1,118 +1,147 @@
+	'gfycat.com',
+	'giphy.com',
+	'reddituploads.com',
+	'redditmedia.com',
+	'twimg.com',
+	'sli.mg',
+	'magaimg.net',
+	'flickr.com',
+	'imgflip.com',
+	'youtube.com',
+	'youtu.be',
+	'youtubedoubler.com',
+	'vimeo.com',
+	'twitch.tv',
+	'streamable.com',
+	'bandcamp.com',
+	'soundcloud.com',
+
+	# not scraper friendly
+
+	'reddit.com',
+	'gyazo.com',
+	'github.com',
+	'xkcd.com',
+	'twitter.com',
+	'spotify.com',
+	'itunes.apple.com',
+	'facebook.com',
+	'gunprime.com',
+	'strawpoll.me',
+	'voyagefusion.com',
+	'rollingstone.com',
+	'google.com',
+	'timeanddate.com',
+	'walmart.com',
+	'roanoke.com',
+	'spotrac.com',
+
+	# original paper excluded wikipedia
+	'wikipedia.org',
+
+	# lots of top posts for this one
+	'battleforthenet.com',
 ])
 
 exclude_extensions = (
-    '.png',
-    '.jpg',
-    '.jpeg',
-    '.gif',
-    '.gifv',
-    '.pdf',
-    '.mp4',
-    '.mp3',
-    '.ogv',
-    '.webm',
-    '.doc',
-    '.docx',
-    '.log',
-    '.csv',
-    '.dat',
-    '.iso',
-    '.bin',
-    '.exe',
-    '.apk',
-    '.jar',
-    '.app',
-    '.ppt',
-    '.pps',
-    '.pptx',
-    '.xml',
-    '.gz',
-    '.xz',
-    '.bz2',
-    '.tgz',
-    '.tar',
-    '.zip',
-    '.wma',
-    '.mov',
-    '.wmv',
-    '.3gp',
-    '.svg',
+	'.png',
+	'.jpg',
+	'.jpeg',
+	'.gif',
+	'.gifv',
+	'.pdf',
+	'.mp4',
+	'.mp3',
+	'.ogv',
+	'.webm',
+	'.doc',
+	'.docx',
+	'.log',
+	'.csv',
+	'.dat',
+	'.iso',
+	'.bin',
+	'.exe',
+	'.apk',
+	'.jar',
+	'.app',
+	'.ppt',
+	'.pps',
+	'.pptx',
+	'.xml',
+	'.gz',
+	'.xz',
+	'.bz2',
+	'.tgz',
+	'.tar',
+	'.zip',
+	'.wma',
+	'.mov',
+	'.wmv',
+	'.3gp',
+	'.svg',
+	'.rar',
+	'.wav',
+	'.avi',
+	'.7z'
 )
 
-
 def should_exclude(url):
-    ext = tldextract.extract(url)
-    domain = '.'.join([x for x in ext if x])
-    basedomain = '.'.join(ext[-2:])
 
-    if basedomain in exclude_domains or domain in exclude_domains:
-        return True
+	ext = tldextract.extract(url)
+	domain = '.'.join([x for x in ext if x])
+	basedomain = '.'.join(ext[-2:])
+
+	# Ignore non-URLs
+	if len(url) <= 8 or ' ' in url or re.match(url_regex, url) is None:
+		return True
+
+	# Ignore excluded domains
+	if basedomain in exclude_domains or domain in exclude_domains:
+		return True
+
+	# Ignore case-insensitive matches for excluded extensions
+	if url.lower().split('?')[0].endswith(exclude_extensions):
+		return True
 
-    if url.split('?')[0].endswith(exclude_extensions):
-        return True
+	return False
 
-    return False
 
 
 if __name__ == '__main__':
-    with open('urls.txt') as urls, open('urls-filtered.txt', 'w') as out:
-        for line in tqdm.tqdm(urls, total=84532938):
-            line = line.strip()
-            if should_exclude(line):
-                continue
+	url_file = 'urls.txt'
+	filtered_file = 'urls-filtered.txt'
 
-            out.write(line + '\n')
+	with open(url_file) as urls, open(filtered_file, 'w') as out:
+		url_len = linecount(url_file)
+		print("URL file is", url_len, "URLs long.")
+		url_set = set()
```

Comment:
How much memory does this use? This should preferably work on low-RAM machines (small cloud instances, etc). Would it be more efficient to sort the url list in-place and then just ignore duplicate consecutive items?

Reply:
Sort needs to have the whole list loaded into memory, so it's generally going to be less space-efficient for this I think; a quick benchmark showed about 4.2 gigabytes of memory for running sort in Bash, versus about 3 for this script's peak. Not sure how much deduplication alone affects the total count of URLs so you may prefer to skip it.

If memory efficiency is critical, the work could be offloaded to the filesystem - base64 each URL and write an empty file with the encoded URL as the filename, then walk the directory and decode all the filenames to recover the unique URLs. Might be easier to just jack up the system's swap space, though.